api/src/main/java/io/grpc/MetricInstrumentRegistry.java

```diff
@@ -16,24 +16,36 @@
 
 package io.grpc;
 
-import java.util.Collections;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import java.util.ArrayList;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
-import java.util.concurrent.CopyOnWriteArrayList;
-import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.atomic.AtomicInteger;
 
 /**
  * A registry for globally registered metric instruments.
  */
 @Internal
 public final class MetricInstrumentRegistry {
+  private static final int DEFAULT_INSTRUMENT_LIST_CAPACITY = 20;
   private static MetricInstrumentRegistry instance;
-  private final List<MetricInstrument> metricInstruments;
+  private final Object lock = new Object();
   private final Set<String> registeredMetricNames;
+  private final AtomicInteger instrumentIndexAlloc = new AtomicInteger();
+  private volatile List<MetricInstrument> metricInstruments;
+  private volatile int instrumentListCapacity = DEFAULT_INSTRUMENT_LIST_CAPACITY;
```

Comment:
This isn't accurate for the tests, because they use the separate constructor. Why doesn't MetricInstrumentRegistryTest use the zero-arg constructor?

Reply:
Makes sense, updated to use zero-arg constructor instead and removed the other test only constructor.